contracts/staking/ValidatorManager.sol

```diff
@@ -93,6 +90,104 @@ abstract contract ValidatorManager is
         $._pChainBlockchainID = settings.pChainBlockchainID;
         $._subnetID = settings.subnetID;
         $._maximumHourlyChurn = settings.maximumHourlyChurn;
+        $._initializedValidatorSet = false;
+    }
+
+    modifier initializedValidatorSet() {
+        require(
+            _getValidatorManagerStorage()._initializedValidatorSet,
+            "ValidatorManager: validator set not initialized"
+        );
+        _;
+    }
+
+    /**
+     * @notice See {IValidatorManager-initializeValidatorSet}.
+     */
+    function initializeValidatorSet(
+        SubnetConversionData calldata subnetConversionData,
+        uint32 messageIndex
+    ) external {
+        ValidatorManagerStorage storage $ = _getValidatorManagerStorage();
+        require(!$._initializedValidatorSet, "ValidatorManager: already initialized validator set");
+
+        WarpMessage memory warpMessage = _getPChainWarpMessage(messageIndex);
+        // Parse the Warp message into SubnetConversionMessage
+        bytes32 subnetConversionID =
+            ValidatorMessages.unpackSubnetConversionMessage(warpMessage.payload);
+
+        // Verify that the sha256 hash of the Subnet conversion data matches with the Warp message's subnetConversionID.
+        bytes memory encodedConversion = abi.encodePacked(
+            subnetConversionData.convertSubnetTxID,
+            subnetConversionData.blockchainID,
+            subnetConversionData.validatorManagerAddress
+        );
+        uint256 length = encodedConversion.length;
+        for (uint256 i; i < length; i++) {
+            InitialValidator memory initialValidator = subnetConversionData.initialValidators[i];
+            encodedConversion = abi.encodePacked(
+                encodedConversion,
+                initialValidator.nodeID,
+                initialValidator.weight,
+                initialValidator.blsPublicKey
+            );
+        }
```

Comment:
This doesn't match up with the serialization in https://github.com/avalanche-foundation/ACPs/pull/146/files. I think the following should work fine?
```suggestion
        bytes memory encodedConversion = abi.encodePacked(subnetConversionData);
```

Reply:
can't pass a struct to encode packed, which part doesn't match up?

Reply:
Ahhh okay... So my understanding (could totally be wrong) is that you need to pack an entire array at once, because the serialization is going to prepend the serialized array with its length, which is needed to deserialize correctly. You'll see that in the ACP, each array has a static 4 bytes in size in addition to the size taken by the fields, which I assume is for the length.

Reply:
ah yes you might be right, need to double check on the additional 4 bytes here

Reply:
updated